Another null pointer exception

Hi,

Using version 0.3 and the following code to search

``` java
QueryBuilder qb3 = QueryBuilders.termQuery("title", queryTitle.toLowerCase());

SearchResults<Listing> list = ElasticSearch.search(qb3, Listing.class);
```

this nullpointerexception is being thrown:

play.exceptions.JavaExecutionException
    at play.mvc.ActionInvoker.invoke(ActionInvoker.java:229)
    at Invocation.HTTP Request(Play!)
Caused by: java.lang.NullPointerException
    at play.modules.elasticsearch.util.ReflectionUtil.setFieldValue(ReflectionUtil.java:351)
    at play.modules.elasticsearch.util.ReflectionUtil.setFieldValue(ReflectionUtil.java:339)
    at play.modules.elasticsearch.transformer.Transformer.toSearchResults(Transformer.java:68)
    at play.modules.elasticsearch.Query.fetch(Query.java:173)
    at play.modules.elasticsearch.ElasticSearch.search(ElasticSearch.java:145)
    at play.modules.elasticsearch.ElasticSearch.search(ElasticSearch.java:96)
    at models.Listing.search(Listing.java:285)
    at controllers.BidAndBuy.search(BidAndBuy.java:15)
    at play.mvc.ActionInvoker.invokeWithContinuation(ActionInvoker.java:546)
    at play.mvc.ActionInvoker.invoke(ActionInvoker.java:500)
    at play.mvc.ActionInvoker.invokeControllerMethod(ActionInvoker.java:476)
    at play.mvc.ActionInvoker.invokeControllerMethod(ActionInvoker.java:471)
    at play.mvc.ActionInvoker.invoke(ActionInvoker.java:159)
    ... 1 more

--- a/src/play/modules/elasticsearch/Query.java
+++ b/src/play/modules/elasticsearch/Query.java
@@ -16,7 +16,8 @@ import play.Logger;
 import play.db.Model;
 import play.modules.elasticsearch.search.SearchResults;
 import play.modules.elasticsearch.transformer.JPATransformer;
-import play.modules.elasticsearch.transformer.Transformer;
+import play.modules.elasticsearch.transformer.MapperTransformer;
+import play.modules.elasticsearch.transformer.SimpleTransformer;
 
 /**
  * An elastic search query
@@ -35,6 +36,7 @@ public class Query<T extends Model> {
 	private int size = -1;
 
 	private boolean hydrate = false;
+	private boolean useMapper = false;
 
 	Query(Class<T> clazz, QueryBuilder builder) {
 		Validate.notNull(clazz, "clazz cannot be null");
@@ -80,6 +82,19 @@ public class Query<T extends Model> {
 	 */
 	public Query<T> hydrate(boolean hydrate) {
 		this.hydrate = hydrate;
+
+		return this;
+	}
+
+	/**
+	 * Controls the usage of mapper
+	 * 
+	 * @param useMapper
+	 *            use mapper during result processing
+	 * @return self
+	 */
+	public Query<T> useMapper(boolean useMapper) {
+		this.useMapper = useMapper;
 
 		return this;
 	}
@@ -168,9 +183,11 @@ public class Query<T extends Model> {
 		SearchResponse searchResponse = request.execute().actionGet();
 		SearchResults<T> searchResults = null;
 		if (hydrate) {
-			searchResults = JPATransformer.toSearchResults(searchResponse, clazz);
+			searchResults = new JPATransformer<T>().toSearchResults(searchResponse, clazz);
+		} else if (useMapper) {
+			searchResults = new MapperTransformer<T>().toSearchResults(searchResponse, clazz);
 		} else {
-			searchResults = Transformer.toSearchResults(searchResponse, clazz);
+			searchResults = new SimpleTransformer<T>().toSearchResults(searchResponse, clazz);
 		}
 		return searchResults;
 	}

--- a/src/play/modules/elasticsearch/adapter/ElasticSearchAdapter.java
+++ b/src/play/modules/elasticsearch/adapter/ElasticSearchAdapter.java
@@ -145,6 +145,7 @@ public abstract class ElasticSearchAdapter {
 
 			contentBuilder = XContentFactory.jsonBuilder().prettyPrint();
 			mapper.addModel(model, contentBuilder);
+			Logger.debug("Index json: %s", contentBuilder.string());
 			IndexResponse response = client.prepareIndex(indexName, typeName, documentId)
 					.setSource(contentBuilder).execute().actionGet();
 

--- a/src/play/modules/elasticsearch/mapping/FieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/FieldMapper.java
@@ -17,30 +17,27 @@ public interface FieldMapper<M> {
 	 * Adds to mapping
 	 * 
 	 * @param builder
-	 * @param prefix
 	 * @throws IOException
 	 */
-	public void addToMapping(XContentBuilder builder, String prefix) throws IOException;
+	public void addToMapping(XContentBuilder builder) throws IOException;
 
 	/**
 	 * Adds to document
 	 * 
 	 * @param model
 	 * @param builder
-	 * @param prefix
 	 * @throws IOException
 	 */
-	public void addToDocument(M model, XContentBuilder builder, String prefix) throws IOException;
+	public void addToDocument(M model, XContentBuilder builder) throws IOException;
 
 	/**
 	 * Inflates a model
 	 * 
 	 * @param model
 	 * @param map
-	 * @param prefix
 	 * @return True if a value was inflated, false otherwise, when no value was
 	 *         present
 	 */
-	public boolean inflate(M model, Map<String, Object> map, String prefix);
+	public boolean inflate(M model, Map<String, Object> map);
 
 }

--- a/src/play/modules/elasticsearch/mapping/FieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/FieldMapper.java
@@ -1,6 +1,7 @@
 package play.modules.elasticsearch.mapping;
 
 import java.io.IOException;
+import java.util.Map;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
@@ -30,5 +31,16 @@ public interface FieldMapper<M> {
 	 * @throws IOException
 	 */
 	public void addToDocument(M model, XContentBuilder builder, String prefix) throws IOException;
+
+	/**
+	 * Inflates a model
+	 * 
+	 * @param model
+	 * @param map
+	 * @param prefix
+	 * @return True if a value was inflated, false otherwise, when no value was
+	 *         present
+	 */
+	public boolean inflate(M model, Map<String, Object> map, String prefix);
 
 }

--- a/src/play/modules/elasticsearch/mapping/MapperFactory.java
+++ b/src/play/modules/elasticsearch/mapping/MapperFactory.java
@@ -52,14 +52,32 @@ public class MapperFactory {
 	 */
 	public static <M> FieldMapper<M> getMapper(Field field) throws MappingException {
 
+		return getMapper(field, null);
+
+	}
+
+	/**
+	 * Gets a {@link FieldMapper} for the specified field, using a prefix in the
+	 * index
+	 * 
+	 * @param <M>
+	 *            the model type
+	 * @param field
+	 *            the field
+	 * @throws MappingException
+	 *             in case of mapping problems
+	 * @return the field mapper
+	 */
+	public static <M> FieldMapper<M> getMapper(Field field, String prefix) throws MappingException {
+
 		if (Collection.class.isAssignableFrom(field.getType())) {
-			return new CollectionFieldMapper<M>(field);
+			return new CollectionFieldMapper<M>(field, prefix);
 
 		} else if (field.isAnnotationPresent(ElasticSearchEmbedded.class)) {
-			return new EmbeddedFieldMapper<M>(field);
+			return new EmbeddedFieldMapper<M>(field, prefix);
 
 		} else {
-			return new SimpleFieldMapper<M>(field);
+			return new SimpleFieldMapper<M>(field, prefix);
 
 		}
 

--- a/src/play/modules/elasticsearch/mapping/MappingUtil.java
+++ b/src/play/modules/elasticsearch/mapping/MappingUtil.java
@@ -5,10 +5,14 @@ import java.lang.annotation.Annotation;
 import java.math.BigDecimal;
 import java.util.Date;
 
+import org.apache.commons.lang.Validate;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
 
 import play.Logger;
+import play.modules.elasticsearch.annotations.ElasticSearchField;
+import play.modules.elasticsearch.annotations.ElasticSearchField.Index;
+import play.modules.elasticsearch.annotations.ElasticSearchField.Store;
 import play.modules.elasticsearch.util.ExceptionUtil;
 
 public abstract class MappingUtil {
@@ -55,6 +59,40 @@ public abstract class MappingUtil {
 		builder.endObject();
 
 		return builder;
+	}
+
+	/**
+	 * Adds a field to the content builder
+	 * 
+	 * @param builder
+	 *            the content builder
+	 * @param name
+	 *            the field name
+	 * @param type
+	 *            the field type
+	 * @param meta
+	 *            the ElasticSearchField annotation (optional) *
+	 * @throws IOException
+	 */
+	public static void addField(XContentBuilder builder, String name, String type,
+			ElasticSearchField meta) throws IOException {
+		Validate.notEmpty(name, "name cannot be empty");
+		Validate.notEmpty(type, "type cannot be empty");
+
+		builder.startObject(name);
+		builder.field("type", type);
+
+		// Check for other settings
+		if (meta != null) {
+			if (meta.index() != Index.NOT_SET) {
+				builder.field("index", meta.index().toString());
+			}
+			if (meta.store() != Store.NOT_SET) {
+				builder.field("store", meta.store().toString());
+			}
+		}
+
+		builder.endObject();
 	}
 
 	/**

--- a/src/play/modules/elasticsearch/mapping/MappingUtil.java
+++ b/src/play/modules/elasticsearch/mapping/MappingUtil.java
@@ -2,6 +2,7 @@ package play.modules.elasticsearch.mapping;
 
 import java.io.IOException;
 import java.lang.annotation.Annotation;
+import java.util.Date;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
@@ -50,5 +51,37 @@ public abstract class MappingUtil {
 		builder.endObject();
 
 		return builder;
+	}
+
+	/**
+	 * Detect the ElasticSearch field type for a {@code Class}
+	 * 
+	 * @param clazz
+	 * @return
+	 */
+	public static String detectFieldType(Class<?> clazz) {
+		// Core types
+		if (String.class.isAssignableFrom(clazz)) {
+			return "string";
+		} else if (Integer.class.isAssignableFrom(clazz) || int.class.isAssignableFrom(clazz)) {
+			return "integer";
+		} else if (Short.class.isAssignableFrom(clazz) || short.class.isAssignableFrom(clazz)) {
+			return "short";
+		} else if (Long.class.isAssignableFrom(clazz) || long.class.isAssignableFrom(clazz)) {
+			return "long";
+		} else if (Float.class.isAssignableFrom(clazz) || float.class.isAssignableFrom(clazz)) {
+			return "float";
+		} else if (Double.class.isAssignableFrom(clazz) || double.class.isAssignableFrom(clazz)) {
+			return "double";
+		} else if (Byte.class.isAssignableFrom(clazz) || byte.class.isAssignableFrom(clazz)) {
+			return "byte";
+		} else if (Date.class.isAssignableFrom(clazz)) {
+			return "date";
+		} else if (Boolean.class.isAssignableFrom(clazz) || boolean.class.isAssignableFrom(clazz)) {
+			return "boolean";
+		}
+
+		// Fall back to string mapping
+		return "string";
 	}
 }

--- a/src/play/modules/elasticsearch/mapping/MappingUtil.java
+++ b/src/play/modules/elasticsearch/mapping/MappingUtil.java
@@ -2,10 +2,14 @@ package play.modules.elasticsearch.mapping;
 
 import java.io.IOException;
 import java.lang.annotation.Annotation;
+import java.math.BigDecimal;
 import java.util.Date;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 import org.elasticsearch.common.xcontent.XContentFactory;
+
+import play.Logger;
+import play.modules.elasticsearch.util.ExceptionUtil;
 
 public abstract class MappingUtil {
 
@@ -83,5 +87,96 @@ public abstract class MappingUtil {
 
 		// Fall back to string mapping
 		return "string";
+	}
+
+	public static Object convertValue(final Object value, final Class<?> targetType) {
+		if (targetType.equals(value.getClass())) {
+			// Types match
+			return value;
+		}
+
+		// Types do not match, perform conversion where needed
+		if (targetType.equals(String.class)) {
+			return value.toString();
+		} else if (targetType.equals(BigDecimal.class)) {
+			return new BigDecimal(value.toString());
+		} else if (targetType.equals(Date.class)) {
+			return convertToDate(value);
+
+			// Use Number intermediary where possible
+		} else if (targetType.equals(Integer.class)) {
+			if (value instanceof Number) {
+				return Integer.valueOf(((Number) value).intValue());
+			} else {
+				return Integer.valueOf(value.toString());
+			}
+		} else if (targetType.equals(Long.class)) {
+			if (value instanceof Number) {
+				return Long.valueOf(((Number) value).longValue());
+			} else {
+				return Long.valueOf(value.toString());
+			}
+		} else if (targetType.equals(Double.class)) {
+			if (value instanceof Number) {
+				return Double.valueOf(((Number) value).doubleValue());
+			} else {
+				return Double.valueOf(value.toString());
+			}
+		} else if (targetType.equals(Float.class)) {
+			if (value instanceof Number) {
+				return Float.valueOf(((Number) value).floatValue());
+			} else {
+				return Float.valueOf(value.toString());
+			}
+
+			// Fallback to simply returning the value
+		} else {
+			return value;
+		}
+	}
+
+	/**
+	 * Convert to date.
+	 * 
+	 * @param value
+	 *            the value
+	 * @return the date
+	 */
+	private static Date convertToDate(Object value) {
+		Date date = null;
+		if (value != null && !"".equals(value)) {
+			if (value instanceof Long) {
+				date = new Date(((Long) value).longValue());
+
+			} else if (value instanceof String) {
+				String val = (String) value;
+				int dateLength = String.valueOf(Long.MAX_VALUE).length();
+				if (dateLength == val.length()) {
+					date = new Date(Long.valueOf(val).longValue());
+				} else {
+					date = getDate(val);
+				}
+			} else {
+				date = (Date) value;
+			}
+		}
+		return date;
+	}
+
+	/**
+	 * Gets the date.
+	 * 
+	 * @param val
+	 *            the val
+	 * @return the date
+	 */
+	private static Date getDate(String val) {
+		try {
+			// Use ES internal converter
+			return XContentBuilder.defaultDatePrinter.parseDateTime(val).toDate();
+		} catch (Throwable t) {
+			Logger.error(ExceptionUtil.getStackTrace(t), val);
+		}
+		return null;
 	}
 }

--- a/src/play/modules/elasticsearch/mapping/ModelMapper.java
+++ b/src/play/modules/elasticsearch/mapping/ModelMapper.java
@@ -1,6 +1,7 @@
 package play.modules.elasticsearch.mapping;
 
 import java.io.IOException;
+import java.util.Map;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
@@ -61,4 +62,12 @@ public interface ModelMapper<M> {
 	 * @throws IOException
 	 */
 	public void addModel(M model, XContentBuilder builder) throws IOException;
+
+	/**
+	 * Inflates a new model from a map of values
+	 * 
+	 * @param map
+	 * @return
+	 */
+	public M createModel(Map<String, Object> map);
 }

--- a/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
@@ -74,11 +74,20 @@ public abstract class AbstractFieldMapper<M> implements FieldMapper<M> {
 	}
 
 	/**
+	 * Gets the field type for the field we represent
+	 * 
+	 * @return
+	 */
+	protected Class<?> getFieldType() {
+		return field.getType();
+	}
+
+	/**
 	 * Gets the ElasticSearch field type for the field we represent
 	 * 
 	 * @return
 	 */
-	protected String getFieldType() {
+	protected String getIndexType() {
 		if (meta != null && meta.type().length() > 0) {
 			// Type was explicitly set, use it
 			return meta.type();

--- a/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
@@ -42,8 +42,7 @@ public abstract class AbstractFieldMapper<M> implements FieldMapper<M> {
 	 *            the content builder
 	 * @throws IOException
 	 */
-	protected static void addField(String name, String type, ElasticSearchField meta,
-			XContentBuilder builder) throws IOException {
+	protected void addField(String name, String type, XContentBuilder builder) throws IOException {
 		// We need at least a type
 		if (type != null) {
 			builder.startObject(name);
@@ -64,7 +63,7 @@ public abstract class AbstractFieldMapper<M> implements FieldMapper<M> {
 		}
 	}
 
-	protected static String detectFieldType(Class<?> clazz) {
+	protected String detectFieldType(Class<?> clazz) {
 		// Core types
 		if (String.class.isAssignableFrom(clazz)) {
 			return "string";

--- a/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
@@ -2,7 +2,6 @@ package play.modules.elasticsearch.mapping.impl;
 
 import java.io.IOException;
 import java.lang.reflect.Field;
-import java.util.Date;
 
 import org.apache.commons.lang.Validate;
 import org.elasticsearch.common.xcontent.XContentBuilder;
@@ -11,6 +10,8 @@ import play.modules.elasticsearch.annotations.ElasticSearchField;
 import play.modules.elasticsearch.annotations.ElasticSearchField.Index;
 import play.modules.elasticsearch.annotations.ElasticSearchField.Store;
 import play.modules.elasticsearch.mapping.FieldMapper;
+import play.modules.elasticsearch.mapping.MappingUtil;
+import play.modules.elasticsearch.util.ReflectionUtil;
 
 /**
  * Abstract base class for {@link FieldMapper}s
@@ -63,30 +64,39 @@ public abstract class AbstractFieldMapper<M> implements FieldMapper<M> {
 		}
 	}
 
-	protected String detectFieldType(Class<?> clazz) {
-		// Core types
-		if (String.class.isAssignableFrom(clazz)) {
-			return "string";
-		} else if (Integer.class.isAssignableFrom(clazz) || int.class.isAssignableFrom(clazz)) {
-			return "integer";
-		} else if (Short.class.isAssignableFrom(clazz) || short.class.isAssignableFrom(clazz)) {
-			return "short";
-		} else if (Long.class.isAssignableFrom(clazz) || long.class.isAssignableFrom(clazz)) {
-			return "long";
-		} else if (Float.class.isAssignableFrom(clazz) || float.class.isAssignableFrom(clazz)) {
-			return "float";
-		} else if (Double.class.isAssignableFrom(clazz) || double.class.isAssignableFrom(clazz)) {
-			return "double";
-		} else if (Byte.class.isAssignableFrom(clazz) || byte.class.isAssignableFrom(clazz)) {
-			return "byte";
-		} else if (Date.class.isAssignableFrom(clazz)) {
-			return "date";
-		} else if (Boolean.class.isAssignableFrom(clazz) || boolean.class.isAssignableFrom(clazz)) {
-			return "boolean";
-		}
+	/**
+	 * Gets the name of the field we represent
+	 * 
+	 * @return
+	 */
+	protected String getFieldName() {
+		return field.getName();
+	}
 
-		// Fall back to string mapping
-		return "string";
+	/**
+	 * Gets the ElasticSearch field type for the field we represent
+	 * 
+	 * @return
+	 */
+	protected String getFieldType() {
+		if (meta != null && meta.type().length() > 0) {
+			// Type was explicitly set, use it
+			return meta.type();
+
+		} else {
+			// Detect type automatically
+			return MappingUtil.detectFieldType(field.getType());
+		}
+	}
+
+	/**
+	 * Gets the value of the field we represent, given a model instance
+	 * 
+	 * @param model
+	 * @return
+	 */
+	protected Object getFieldValue(M model) {
+		return ReflectionUtil.getFieldValue(model, field);
 	}
 
 }

--- a/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/AbstractFieldMapper.java
@@ -1,14 +1,10 @@
 package play.modules.elasticsearch.mapping.impl;
 
-import java.io.IOException;
 import java.lang.reflect.Field;
 
 import org.apache.commons.lang.Validate;
-import org.elasticsearch.common.xcontent.XContentBuilder;
 
 import play.modules.elasticsearch.annotations.ElasticSearchField;
-import play.modules.elasticsearch.annotations.ElasticSearchField.Index;
-import play.modules.elasticsearch.annotations.ElasticSearchField.Store;
 import play.modules.elasticsearch.mapping.FieldMapper;
 import play.modules.elasticsearch.mapping.MappingUtil;
 import play.modules.elasticsearch.util.ReflectionUtil;
@@ -23,44 +19,39 @@ public abstract class AbstractFieldMapper<M> implements FieldMapper<M> {
 
 	protected final Field field;
 	protected final ElasticSearchField meta;
+	private final String prefix, indexField;
 
-	public AbstractFieldMapper(Field field) {
+	public AbstractFieldMapper(Field field, String prefix) {
 		Validate.notNull(field, "field cannot be null");
 		this.field = field;
 		this.meta = field.getAnnotation(ElasticSearchField.class);
+		this.prefix = prefix;
+
+		// Maybe this a premature optimization, but getIndexField() will be
+		// called a lot
+		indexField = prefix(field.getName());
 	}
 
 	/**
-	 * Adds a field to the content builder
+	 * Gets the prefix to use when indexing this field
 	 * 
-	 * @param name
-	 *            the field name
-	 * @param type
-	 *            the field type
-	 * @param meta
-	 *            the ElasticSearchField annotation (optional)
-	 * @param builder
-	 *            the content builder
-	 * @throws IOException
+	 * @return
 	 */
-	protected void addField(String name, String type, XContentBuilder builder) throws IOException {
-		// We need at least a type
-		if (type != null) {
-			builder.startObject(name);
+	protected String getPrefix() {
+		return prefix;
+	}
 
-			builder.field("type", type);
-
-			// Check for other settings
-			if (meta != null) {
-				if (meta.index() != Index.NOT_SET) {
-					builder.field("index", meta.index().toString());
-				}
-				if (meta.store() != Store.NOT_SET) {
-					builder.field("store", meta.store().toString());
-				}
-			}
-
-			builder.endObject();
+	/**
+	 * Prefixes a value with our prefix, if we have one
+	 * 
+	 * @param value
+	 * @return
+	 */
+	protected String prefix(String value) {
+		if (prefix != null) {
+			return prefix + value;
+		} else {
+			return value;
 		}
 	}
 
@@ -80,6 +71,15 @@ public abstract class AbstractFieldMapper<M> implements FieldMapper<M> {
 	 */
 	protected Class<?> getFieldType() {
 		return field.getType();
+	}
+
+	/**
+	 * Gets the field we should use in the index
+	 * 
+	 * @return
+	 */
+	protected String getIndexField() {
+		return indexField;
 	}
 
 	/**

--- a/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
@@ -83,7 +83,7 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 	@Override
 	public void addToDocument(M model, XContentBuilder builder) throws IOException {
 		String indexFieldName = getIndexField();
-		Collection<?> value = (Collection<?>) ReflectionUtil.getFieldValue(model, field);
+		Collection<?> value = (Collection<?>) getFieldValue(model);
 
 		if (value != null) {
 			builder.startArray(indexFieldName);

--- a/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
@@ -13,6 +13,7 @@ import play.modules.elasticsearch.annotations.ElasticSearchEmbedded;
 import play.modules.elasticsearch.mapping.FieldMapper;
 import play.modules.elasticsearch.mapping.MapperFactory;
 import play.modules.elasticsearch.mapping.MappingException;
+import play.modules.elasticsearch.mapping.MappingUtil;
 import play.modules.elasticsearch.util.ReflectionUtil;
 
 /**
@@ -41,14 +42,14 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 		fields = new ArrayList<FieldMapper<Object>>();
 		if (nestedMode) {
 			Class<?> itemClass = getCollectionType();
-			type = detectFieldType(itemClass);
+			type = MappingUtil.detectFieldType(itemClass);
 			List<Field> fieldsToIndex = EmbeddedFieldMapper.getFieldsToIndex(itemClass, embed);
 
 			for (Field embeddedField : fieldsToIndex) {
 				fields.add(MapperFactory.getMapper(embeddedField));
 			}
 		} else {
-			type = detectFieldType(getCollectionType());
+			type = MappingUtil.detectFieldType(getCollectionType());
 		}
 	}
 

--- a/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
@@ -6,6 +6,7 @@ import java.lang.reflect.ParameterizedType;
 import java.util.ArrayList;
 import java.util.Collection;
 import java.util.List;
+import java.util.Map;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
@@ -107,6 +108,49 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 			}
 
 			builder.endArray();
+		}
+	}
+
+	@Override
+	public boolean inflate(M model, Map<String, Object> map, String prefix) {
+		String name = getFieldName();
+		final List<Object> input = (List<Object>) map.get(name);
+		final Collection<Object> output = (Collection<Object>) getFieldValue(model);
+		final Class<?> type = getCollectionType();
+
+		// If we have input and output, continue
+		if (input != null && output != null) {
+			if (nestedMode) {
+				// Embedded mode uses mapping
+				for (Object inputItem : input) {
+					// Fetch input item fields
+					Map<String, Object> inputItemFields = (Map<String, Object>) inputItem;
+
+					// Create new target instance
+					Object outputItem = ReflectionUtil.newInstance(type);
+
+					for (FieldMapper<Object> mapper : fields) {
+						mapper.inflate(outputItem, inputItemFields, null);
+					}
+
+					output.add(outputItem);
+				}
+			} else {
+				// Flat mode uses primitive values or toString
+				for (Object inputItem : input) {
+					// Try to convert
+					Object outputItem = MappingUtil.convertValue(inputItem, type);
+
+					// This should only succeed for simple types
+					if (type.isAssignableFrom(outputItem.getClass())) {
+						output.add(outputItem);
+					}
+				}
+			}
+
+			return true;
+		} else {
+			return false;
 		}
 	}
 

--- a/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
@@ -70,7 +70,7 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 			builder.endObject();
 		} else {
 			// Flat mode (array of primitives)
-			addField(field.getName(), type, meta, builder);
+			addField(field.getName(), type, builder);
 		}
 	}
 

--- a/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/CollectionFieldMapper.java
@@ -29,8 +29,8 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 	private final String type;
 	private final List<FieldMapper<Object>> fields;
 
-	public CollectionFieldMapper(Field field) {
-		super(field);
+	public CollectionFieldMapper(Field field, String prefix) {
+		super(field, prefix);
 
 		if (!Collection.class.isAssignableFrom(field.getType())) {
 			throw new MappingException("field must be of Collection type");
@@ -39,18 +39,20 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 		ElasticSearchEmbedded embed = field.getAnnotation(ElasticSearchEmbedded.class);
 		nestedMode = (embed != null);
 
+		// Detect object type in collection
+		type = MappingUtil.detectFieldType(getCollectionType());
+
 		// Find fields to use for embedded objects
-		fields = new ArrayList<FieldMapper<Object>>();
 		if (nestedMode) {
 			Class<?> itemClass = getCollectionType();
-			type = MappingUtil.detectFieldType(itemClass);
 			List<Field> fieldsToIndex = EmbeddedFieldMapper.getFieldsToIndex(itemClass, embed);
+			fields = new ArrayList<FieldMapper<Object>>();
 
 			for (Field embeddedField : fieldsToIndex) {
 				fields.add(MapperFactory.getMapper(embeddedField));
 			}
 		} else {
-			type = MappingUtil.detectFieldType(getCollectionType());
+			fields = null;
 		}
 	}
 
@@ -60,36 +62,38 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 	}
 
 	@Override
-	public void addToMapping(XContentBuilder builder, String prefix) throws IOException {
+	public void addToMapping(XContentBuilder builder) throws IOException {
+		String indexFieldName = getIndexField();
+
 		if (nestedMode) {
 			// Embedded mode
-			builder.startObject(field.getName());
+			builder.startObject(indexFieldName);
 			builder.startObject("properties");
 			for (FieldMapper<?> mapper : fields) {
-				mapper.addToMapping(builder, null);
+				mapper.addToMapping(builder);
 			}
 			builder.endObject();
 			builder.endObject();
 		} else {
 			// Flat mode (array of primitives)
-			addField(field.getName(), type, builder);
+			MappingUtil.addField(builder, indexFieldName, type, meta);
 		}
 	}
 
 	@Override
-	public void addToDocument(M model, XContentBuilder builder, String prefix) throws IOException {
-		String name = field.getName();
+	public void addToDocument(M model, XContentBuilder builder) throws IOException {
+		String indexFieldName = getIndexField();
 		Collection<?> value = (Collection<?>) ReflectionUtil.getFieldValue(model, field);
 
 		if (value != null) {
-			builder.startArray(name);
+			builder.startArray(indexFieldName);
 
 			if (nestedMode) {
 				// Embedded mode uses mapping
 				for (Object object : (Collection<?>) value) {
 					builder.startObject();
 					for (FieldMapper<Object> mapper : fields) {
-						mapper.addToDocument(object, builder, null);
+						mapper.addToDocument(object, builder);
 					}
 					builder.endObject();
 				}
@@ -112,38 +116,38 @@ public class CollectionFieldMapper<M> extends AbstractFieldMapper<M> {
 	}
 
 	@Override
-	public boolean inflate(M model, Map<String, Object> map, String prefix) {
-		String name = getFieldName();
-		final List<Object> input = (List<Object>) map.get(name);
-		final Collection<Object> output = (Collection<Object>) getFieldValue(model);
+	public boolean inflate(M model, Map<String, Object> map) {
+		String indexFieldName = getIndexField();
+		final List<Object> indexValue = (List<Object>) map.get(indexFieldName);
+		final Collection<Object> modelValue = (Collection<Object>) getFieldValue(model);
 		final Class<?> type = getCollectionType();
 
 		// If we have input and output, continue
-		if (input != null && output != null) {
+		if (indexValue != null && modelValue != null) {
 			if (nestedMode) {
 				// Embedded mode uses mapping
-				for (Object inputItem : input) {
+				for (Object indexItem : indexValue) {
 					// Fetch input item fields
-					Map<String, Object> inputItemFields = (Map<String, Object>) inputItem;
+					Map<String, Object> indexItemMap = (Map<String, Object>) indexItem;
 
 					// Create new target instance
 					Object outputItem = ReflectionUtil.newInstance(type);
 
 					for (FieldMapper<Object> mapper : fields) {
-						mapper.inflate(outputItem, inputItemFields, null);
+						mapper.inflate(outputItem, indexItemMap);
 					}
 
-					output.add(outputItem);
+					modelValue.add(outputItem);
 				}
 			} else {
 				// Flat mode uses primitive values or toString
-				for (Object inputItem : input) {
+				for (Object indexItem : indexValue) {
 					// Try to convert
-					Object outputItem = MappingUtil.convertValue(inputItem, type);
+					Object modelItem = MappingUtil.convertValue(indexItem, type);
 
 					// This should only succeed for simple types
-					if (type.isAssignableFrom(outputItem.getClass())) {
-						output.add(outputItem);
+					if (type.isAssignableFrom(modelItem.getClass())) {
+						modelValue.add(modelItem);
 					}
 				}
 			}

--- a/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
@@ -124,7 +124,7 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 	@Override
 	public void addToDocument(M model, XContentBuilder builder) throws IOException {
 		String name = getIndexField();
-		Object value = ReflectionUtil.getFieldValue(model, field);
+		Object value = getFieldValue(model);
 
 		if (value != null) {
 			switch (embed.mode()) {

--- a/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
@@ -9,6 +9,7 @@ import java.util.Map;
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
 import play.modules.elasticsearch.annotations.ElasticSearchEmbedded;
+import play.modules.elasticsearch.annotations.ElasticSearchEmbedded.Mode;
 import play.modules.elasticsearch.mapping.FieldMapper;
 import play.modules.elasticsearch.mapping.MapperFactory;
 import play.modules.elasticsearch.mapping.MappingException;
@@ -24,25 +25,27 @@ import scala.actors.threadpool.Arrays;
 public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 
 	private final ElasticSearchEmbedded embed;
-	private final String prefix;
 	private final List<FieldMapper<Object>> fields;
 
-	public EmbeddedFieldMapper(Field field) {
-		super(field);
+	public EmbeddedFieldMapper(Field field, String prefix) {
+		super(field, prefix);
 
 		embed = field.getAnnotation(ElasticSearchEmbedded.class);
 
-		// Check for prefix
-		if (embed.prefix().length() > 0) {
-			this.prefix = embed.prefix();
-		} else {
-			this.prefix = field.getName() + ".";
+		// Set correct prefix in case we are in embedded mode
+		String embedPrefix = null;
+		if (embed.mode() == Mode.embedded) {
+			if (embed.prefix().length() > 0) {
+				embedPrefix = prefix(embed.prefix());
+			} else {
+				embedPrefix = getFieldName() + ".";
+			}
 		}
 
 		// Add fieldmappers for embedded fields
 		fields = new ArrayList<FieldMapper<Object>>();
 		for (Field embeddedField : getFieldsToIndex(field.getType(), embed)) {
-			fields.add(MapperFactory.getMapper(embeddedField));
+			fields.add(MapperFactory.getMapper(embeddedField, embedPrefix));
 		}
 	}
 
@@ -95,23 +98,22 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 	}
 
 	@Override
-	public void addToMapping(XContentBuilder builder, String prefix) throws IOException {
-		String name = field.getName();
+	public void addToMapping(XContentBuilder builder) throws IOException {
+		String indexFieldName = getIndexField();
 
 		switch (embed.mode()) {
 		case embedded:
-			String prefixToUse = (prefix != null) ? prefix + this.prefix : this.prefix;
 			for (FieldMapper<?> mapper : fields) {
-				mapper.addToMapping(builder, prefixToUse);
+				mapper.addToMapping(builder);
 			}
 			break;
 		case object:
 		case nested:
-			builder.startObject(name);
+			builder.startObject(indexFieldName);
 			builder.field("type", embed.mode().toString());
 			builder.startObject("properties");
 			for (FieldMapper<?> mapper : fields) {
-				mapper.addToMapping(builder, null);
+				mapper.addToMapping(builder);
 			}
 			builder.endObject();
 			builder.endObject();
@@ -120,23 +122,22 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 	}
 
 	@Override
-	public void addToDocument(M model, XContentBuilder builder, String prefix) throws IOException {
-		String name = field.getName();
+	public void addToDocument(M model, XContentBuilder builder) throws IOException {
+		String name = getIndexField();
 		Object value = ReflectionUtil.getFieldValue(model, field);
 
 		if (value != null) {
 			switch (embed.mode()) {
 			case embedded:
-				String prefixToUse = (prefix != null) ? prefix + this.prefix : this.prefix;
 				for (FieldMapper<Object> mapper : fields) {
-					mapper.addToDocument(value, builder, prefixToUse);
+					mapper.addToDocument(value, builder);
 				}
 				break;
 			case object:
 			case nested:
 				builder.startObject(name);
 				for (FieldMapper<Object> mapper : fields) {
-					mapper.addToDocument(value, builder, null);
+					mapper.addToDocument(value, builder);
 				}
 				builder.endObject();
 				break;
@@ -145,7 +146,7 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 	}
 
 	@Override
-	public boolean inflate(M model, Map<String, Object> map, String prefix) {
+	public boolean inflate(M model, Map<String, Object> map) {
 		String name = getFieldName();
 
 		// Create new target instance
@@ -156,9 +157,8 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 
 		switch (embed.mode()) {
 		case embedded:
-			String prefixToUse = (prefix != null) ? prefix + this.prefix : this.prefix;
 			for (FieldMapper<Object> mapper : fields) {
-				if (mapper.inflate(value, map, prefixToUse)) {
+				if (mapper.inflate(value, map)) {
 					nonNullValue = true;
 				}
 			}
@@ -170,7 +170,7 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 				Map<String, Object> nestedMap = (Map<String, Object>) input;
 
 				for (FieldMapper<Object> mapper : fields) {
-					mapper.inflate(value, nestedMap, null);
+					mapper.inflate(value, nestedMap);
 				}
 
 				nonNullValue = true;

--- a/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
@@ -4,6 +4,7 @@ import java.io.IOException;
 import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
@@ -140,6 +141,48 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 				builder.endObject();
 				break;
 			}
+		}
+	}
+
+	@Override
+	public boolean inflate(M model, Map<String, Object> map, String prefix) {
+		String name = getFieldName();
+
+		// Create new target instance
+		Object value = ReflectionUtil.newInstance(getFieldType());
+
+		// Keep track if we found any field (indicator of non-null output)
+		boolean nonNullValue = false;
+
+		switch (embed.mode()) {
+		case embedded:
+			String prefixToUse = (prefix != null) ? prefix + this.prefix : this.prefix;
+			for (FieldMapper<Object> mapper : fields) {
+				if (mapper.inflate(value, map, prefixToUse)) {
+					nonNullValue = true;
+				}
+			}
+			break;
+		case object:
+		case nested:
+			Object input = map.get(name);
+			if (input != null) {
+				Map<String, Object> nestedMap = (Map<String, Object>) input;
+
+				for (FieldMapper<Object> mapper : fields) {
+					mapper.inflate(value, nestedMap, null);
+				}
+
+				nonNullValue = true;
+			}
+			break;
+		}
+
+		if (nonNullValue) {
+			ReflectionUtil.setFieldValue(model, name, value);
+			return true;
+		} else {
+			return false;
 		}
 	}
 

--- a/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/EmbeddedFieldMapper.java
@@ -10,6 +10,7 @@ import org.elasticsearch.common.xcontent.XContentBuilder;
 import play.modules.elasticsearch.annotations.ElasticSearchEmbedded;
 import play.modules.elasticsearch.mapping.FieldMapper;
 import play.modules.elasticsearch.mapping.MapperFactory;
+import play.modules.elasticsearch.mapping.MappingException;
 import play.modules.elasticsearch.util.ReflectionUtil;
 import scala.actors.threadpool.Arrays;
 
@@ -47,10 +48,28 @@ public class EmbeddedFieldMapper<M> extends AbstractFieldMapper<M> {
 	static List<Field> getFieldsToIndex(Class<?> clazz, ElasticSearchEmbedded meta) {
 		@SuppressWarnings("unchecked")
 		List<String> fieldsToIndex = Arrays.asList(meta.fields());
+		List<Field> clazzFields = ReflectionUtil.getAllFields(clazz);
 		List<Field> fields = new ArrayList<Field>();
 
+		// Make sure the user has not requested unknown fields
+		if (fieldsToIndex.size() > 0) {
+			for (String fieldName : fieldsToIndex) {
+				boolean knownField = false;
+				for (Field clazzField : clazzFields) {
+					if (clazzField.getName().equals(fieldName)) {
+						knownField = true;
+						break;
+					}
+				}
+
+				if (!knownField) {
+					throw new MappingException("Unknown field specified in " + meta);
+				}
+			}
+		}
+
 		// Set up fields
-		for (Field embeddedField : ReflectionUtil.getAllFields(clazz)) {
+		for (Field embeddedField : clazzFields) {
 			if (PlayModelMapper.shouldIgnoreField(embeddedField)) {
 				continue;
 			}

--- a/src/play/modules/elasticsearch/mapping/impl/PlayModelMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/PlayModelMapper.java
@@ -4,6 +4,7 @@ import java.io.IOException;
 import java.lang.reflect.Field;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Map;
 
 import javax.persistence.Transient;
 
@@ -145,6 +146,17 @@ public class PlayModelMapper<M extends Model> implements ModelMapper<M> {
 		}
 
 		builder.endObject();
+	}
+
+	@Override
+	public M createModel(Map<String, Object> map) {
+		M model = ReflectionUtil.newInstance(clazz);
+
+		for (FieldMapper<M> field : mapping) {
+			field.inflate(model, map, null);
+		}
+
+		return model;
 	}
 
 }

--- a/src/play/modules/elasticsearch/mapping/impl/PlayModelMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/PlayModelMapper.java
@@ -130,7 +130,7 @@ public class PlayModelMapper<M extends Model> implements ModelMapper<M> {
 		builder.startObject("properties");
 
 		for (FieldMapper<M> field : mapping) {
-			field.addToMapping(builder, null);
+			field.addToMapping(builder);
 		}
 
 		builder.endObject();
@@ -142,7 +142,7 @@ public class PlayModelMapper<M extends Model> implements ModelMapper<M> {
 		builder.startObject();
 
 		for (FieldMapper<M> field : mapping) {
-			field.addToDocument(model, builder, null);
+			field.addToDocument(model, builder);
 		}
 
 		builder.endObject();
@@ -153,7 +153,7 @@ public class PlayModelMapper<M extends Model> implements ModelMapper<M> {
 		M model = ReflectionUtil.newInstance(clazz);
 
 		for (FieldMapper<M> field : mapping) {
-			field.inflate(model, map, null);
+			field.inflate(model, map);
 		}
 
 		return model;

--- a/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
@@ -5,8 +5,6 @@ import java.lang.reflect.Field;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
-import play.modules.elasticsearch.util.ReflectionUtil;
-
 /**
  * Field mapper for simple, single-valued types
  * 
@@ -21,7 +19,7 @@ public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 
 	@Override
 	public void addToMapping(XContentBuilder builder, String prefix) throws IOException {
-		String name = field.getName();
+		String name = getFieldName();
 		String type = getFieldType();
 
 		if (prefix != null) {
@@ -33,8 +31,8 @@ public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 
 	@Override
 	public void addToDocument(M model, XContentBuilder builder, String prefix) throws IOException {
-		String name = field.getName();
-		Object value = ReflectionUtil.getFieldValue(model, field);
+		String name = getFieldName();
+		Object value = getFieldValue(model);
 
 		if (value != null) {
 			if (prefix != null) {
@@ -42,17 +40,6 @@ public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 			} else {
 				builder.field(name, value);
 			}
-		}
-	}
-
-	protected String getFieldType() {
-		if (meta != null && meta.type().length() > 0) {
-			// Type was explicitly set, use it
-			return meta.type();
-
-		} else {
-			// Detect type automatically
-			return detectFieldType(field.getType());
 		}
 	}
 

--- a/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
@@ -25,9 +25,9 @@ public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 		String type = getFieldType();
 
 		if (prefix != null) {
-			addField(prefix + name, type, meta, builder);
+			addField(prefix + name, type, builder);
 		} else {
-			addField(name, type, meta, builder);
+			addField(name, type, builder);
 		}
 	}
 

--- a/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
@@ -6,6 +6,7 @@ import java.util.Map;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
 
+import play.modules.elasticsearch.mapping.MappingUtil;
 import play.modules.elasticsearch.util.ReflectionUtil;
 
 /**
@@ -16,45 +17,37 @@ import play.modules.elasticsearch.util.ReflectionUtil;
  */
 public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 
-	public SimpleFieldMapper(Field field) {
-		super(field);
+	public SimpleFieldMapper(Field field, String prefix) {
+		super(field, prefix);
 	}
 
 	@Override
-	public void addToMapping(XContentBuilder builder, String prefix) throws IOException {
-		String name = getFieldName();
+	public void addToMapping(XContentBuilder builder) throws IOException {
+		String field = getIndexField();
 		String type = getIndexType();
 
-		if (prefix != null) {
-			addField(prefix + name, type, builder);
-		} else {
-			addField(name, type, builder);
-		}
+		MappingUtil.addField(builder, field, type, meta);
 	}
 
 	@Override
-	public void addToDocument(M model, XContentBuilder builder, String prefix) throws IOException {
-		String name = getFieldName();
+	public void addToDocument(M model, XContentBuilder builder) throws IOException {
+		String field = getIndexField();
 		Object value = getFieldValue(model);
 
 		if (value != null) {
-			if (prefix != null) {
-				builder.field(prefix + name, value);
-			} else {
-				builder.field(name, value);
-			}
+			builder.field(field, value);
 		}
 	}
 
 	@Override
-	public boolean inflate(M model, Map<String, Object> map, String prefix) {
-		String name = getFieldName();
-		String indexName = (prefix != null) ? prefix + name : name;
+	public boolean inflate(M model, Map<String, Object> map) {
+		String modelFieldName = getFieldName();
+		String indexFieldName = getIndexField();
 
-		if (map.containsKey(name)) {
-			Object value = map.get(indexName);
+		if (map.containsKey(indexFieldName)) {
+			Object value = map.get(indexFieldName);
 			if (value != null) {
-				ReflectionUtil.setFieldValue(model, name, value);
+				ReflectionUtil.setFieldValue(model, modelFieldName, value);
 				return true;
 			}
 		}

--- a/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
+++ b/src/play/modules/elasticsearch/mapping/impl/SimpleFieldMapper.java
@@ -2,8 +2,11 @@ package play.modules.elasticsearch.mapping.impl;
 
 import java.io.IOException;
 import java.lang.reflect.Field;
+import java.util.Map;
 
 import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import play.modules.elasticsearch.util.ReflectionUtil;
 
 /**
  * Field mapper for simple, single-valued types
@@ -20,7 +23,7 @@ public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 	@Override
 	public void addToMapping(XContentBuilder builder, String prefix) throws IOException {
 		String name = getFieldName();
-		String type = getFieldType();
+		String type = getIndexType();
 
 		if (prefix != null) {
 			addField(prefix + name, type, builder);
@@ -41,6 +44,22 @@ public class SimpleFieldMapper<M> extends AbstractFieldMapper<M> {
 				builder.field(name, value);
 			}
 		}
+	}
+
+	@Override
+	public boolean inflate(M model, Map<String, Object> map, String prefix) {
+		String name = getFieldName();
+		String indexName = (prefix != null) ? prefix + name : name;
+
+		if (map.containsKey(name)) {
+			Object value = map.get(indexName);
+			if (value != null) {
+				ReflectionUtil.setFieldValue(model, name, value);
+				return true;
+			}
+		}
+
+		return false;
 	}
 
 }

--- a/src/play/modules/elasticsearch/transformer/JPATransformer.java
+++ b/src/play/modules/elasticsearch/transformer/JPATransformer.java
@@ -19,10 +19,10 @@ import play.modules.elasticsearch.search.SearchResults;
  * Transforms ES SearchResponse to a list of hydrated entities
  * 
  * @author Bas
- *
+ * 
  */
-public class JPATransformer {
-	
+public class JPATransformer<T extends Model> implements Transformer<T> {
+
 	/**
 	 * To search results.
 	 * 
@@ -34,8 +34,7 @@ public class JPATransformer {
 	 *            the clazz
 	 * @return the search results
 	 */
-	public static <T extends Model> SearchResults<T> toSearchResults(
-			SearchResponse searchResponse, Class<T> clazz) {
+	public SearchResults<T> toSearchResults(SearchResponse searchResponse, Class<T> clazz) {
 		// Get Total Records Found
 		long count = searchResponse.hits().totalHits();
 
@@ -48,28 +47,30 @@ public class JPATransformer {
 		for (SearchHit h : searchResponse.hits()) {
 			try {
 				ids.add(Binder.directBind(h.getId(), keyType));
-	        } catch (Exception e) {
-	            throw new UnexpectedException("Could not convert the ID from index to corresponding type", e);
-	        }
+			} catch (Exception e) {
+				throw new UnexpectedException(
+						"Could not convert the ID from index to corresponding type", e);
+			}
 		}
-		
+
 		Logger.debug("Model IDs returned by ES: %s", ids);
 
 		List<T> objects = null;
-		
-		if( ids.size() > 0 ) {
+
+		if (ids.size() > 0) {
 			// Fetch JPA entities from database while preserving ES result order
 			objects = loadFromDb(clazz, ids);
 			sortByIds(objects, ids);
-			
+
 			// Make sure all items exist in the database
-			if( objects.size() != ids.size() ) {
-				throw new IllegalStateException("Please re-index, not all indexed items are available in the database");
+			if (objects.size() != ids.size()) {
+				throw new IllegalStateException(
+						"Please re-index, not all indexed items are available in the database");
 			}
 		} else {
 			objects = Collections.emptyList();
 		}
-		
+
 		Logger.debug("Models after sorting: %s", objects);
 
 		// Return Results
@@ -86,14 +87,15 @@ public class JPATransformer {
 	 */
 	private static <T extends Model> List<T> loadFromDb(Class<T> clazz, List<Object> ids) {
 		// JPA maps the "id" field to the key automatically
-		List<T> objects = JPQL.instance.find(clazz.getName(), "id in (?1)",
-				new Object[] { ids }).fetch();
+		List<T> objects = JPQL.instance.find(clazz.getName(), "id in (?1)", new Object[] { ids })
+				.fetch();
 
 		return objects;
 	}
 
 	/**
-	 * Sort list of objects according to the order of their keys as defined by ids
+	 * Sort list of objects according to the order of their keys as defined by
+	 * ids
 	 * 
 	 * @param <T>
 	 * @param objects

--- a/src/play/modules/elasticsearch/transformer/SimpleTransformer.java
+++ b/src/play/modules/elasticsearch/transformer/SimpleTransformer.java
@@ -1,0 +1,87 @@
+/** 
+ * Copyright 2011 The Apache Software Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * 
+ * @author Felipe Oliveira (http://mashup.fm)
+ * 
+ */
+package play.modules.elasticsearch.transformer;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.search.SearchHit;
+
+import play.Logger;
+import play.db.Model;
+import play.modules.elasticsearch.search.SearchResults;
+import play.modules.elasticsearch.util.ReflectionUtil;
+
+/**
+ * The original transformer, as written by Felipe.
+ * 
+ * <p>
+ * This transformer only supports basic mappings, without nested collections or
+ * models.
+ * 
+ * @param <T>
+ *            the generic type
+ */
+public class SimpleTransformer<T extends Model> implements Transformer<T> {
+
+	/**
+	 * To search results.
+	 * 
+	 * @param <T>
+	 *            the generic type
+	 * @param searchResponse
+	 *            the search response
+	 * @param clazz
+	 *            the clazz
+	 * @return the search results
+	 */
+	public SearchResults<T> toSearchResults(SearchResponse searchResponse, Class<T> clazz) {
+		// Get Total Records Found
+		long count = searchResponse.hits().totalHits();
+
+		// Init List
+		List<T> objects = new ArrayList<T>();
+
+		// Loop on each one
+		for (SearchHit h : searchResponse.hits()) {
+			// Init Model Class
+			Logger.debug("Starting Record!");
+			T o = ReflectionUtil.newInstance(clazz);
+
+			// Get Data Map
+			Map<String, Object> map = h.sourceAsMap();
+			Logger.debug("Record Map: %s", map);
+
+			// Bind Data
+			for (Map.Entry<String, Object> e : map.entrySet()) {
+				ReflectionUtil.setFieldValue(o, e.getKey(), e.getValue());
+			}
+
+			// Log Debug
+			Logger.debug("Model Instance: %s", o);
+			objects.add(o);
+		}
+
+		// Return Results
+		return new SearchResults<T>(count, objects, searchResponse.facets());
+	}
+
+}

--- a/src/play/modules/elasticsearch/transformer/Transformer.java
+++ b/src/play/modules/elasticsearch/transformer/Transformer.java
@@ -1,80 +1,25 @@
-/** 
- * Copyright 2011 The Apache Software Foundation
- *
- * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
- * You may obtain a copy of the License at
- *
- *     http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and
- * limitations under the License.
- * 
- * @author Felipe Oliveira (http://mashup.fm)
- * 
- */
 package play.modules.elasticsearch.transformer;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
-
-import play.modules.elasticsearch.search.SearchResults;
-import play.modules.elasticsearch.util.ReflectionUtil;
-
 import org.elasticsearch.action.search.SearchResponse;
-import org.elasticsearch.search.SearchHit;
 
-import play.Logger;
 import play.db.Model;
+import play.modules.elasticsearch.search.SearchResults;
 
 /**
- * The Class Transformer.
- *
- * @param <T> the generic type
+ * Transformer for search results
+ * 
+ * @author Bas
+ * 
+ * @param <T>
  */
-public class Transformer<T extends Model> {
+public interface Transformer<T extends Model> {
 
 	/**
-	 * To search results.
-	 *
-	 * @param <T> the generic type
-	 * @param searchResponse the search response
-	 * @param clazz the clazz
-	 * @return the search results
+	 * Transforms a {@link SearchResponse} into {@link SearchResults}
+	 * 
+	 * @param searchResponse
+	 * @param clazz
+	 * @return
 	 */
-	public static <T extends Model> SearchResults<T> toSearchResults(SearchResponse searchResponse, Class<T> clazz) {
-		// Get Total Records Found
-		long count = searchResponse.hits().totalHits();
-
-		// Init List
-		List<T> objects = new ArrayList<T>();
-
-		// Loop on each one
-		for (SearchHit h : searchResponse.hits()) {
-			// Init Model Class
-			Logger.debug("Starting Record!");
-			T o = ReflectionUtil.newInstance(clazz);
-
-			// Get Data Map
-			Map<String, Object> map = h.sourceAsMap();
-			Logger.debug("Record Map: %s", map);
-
-			// Bind Data
-			for (Map.Entry<String, Object> e : map.entrySet()) {
-				ReflectionUtil.setFieldValue(o, e.getKey(), e.getValue());
-			}
-
-			// Log Debug
-			Logger.debug("Model Instance: %s", o);
-			objects.add(o);
-		}
-
-		// Return Results
-		return new SearchResults<T>(count, objects, searchResponse.facets());
-	}
-
+	public SearchResults<T> toSearchResults(SearchResponse searchResponse, Class<T> clazz);
 }

--- a/src/play/modules/elasticsearch/util/ReflectionUtil.java
+++ b/src/play/modules/elasticsearch/util/ReflectionUtil.java
@@ -19,21 +19,19 @@
 package play.modules.elasticsearch.util;
 
 import java.lang.annotation.Annotation;
+import java.lang.reflect.Constructor;
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
 import java.lang.reflect.Type;
-import java.math.BigDecimal;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
-import java.util.Date;
 import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
-import org.elasticsearch.common.xcontent.XContentBuilder;
-
 import play.Logger;
+import play.modules.elasticsearch.mapping.MappingUtil;
 
 /**
  * The Class ReflectionUtil.
@@ -46,6 +44,9 @@ public abstract class ReflectionUtil {
 	/** The Constant classFieldsCache. */
 	private static final ConcurrentMap<String, List<Field>> classFieldsCache = new ConcurrentHashMap<String, List<Field>>();
 
+	/** Constructor cache */
+	private static final ConcurrentMap<Class<?>, Constructor<?>> classConstructorCache = new ConcurrentHashMap<Class<?>, Constructor<?>>();
+
 	/**
 	 * Instantiates a new reflection util.
 	 */
@@ -56,6 +57,7 @@ public abstract class ReflectionUtil {
 	public static void clearCache() {
 		annotationFieldsCache.clear();
 		classFieldsCache.clear();
+		classConstructorCache.clear();
 	}
 
 	/**
@@ -222,12 +224,8 @@ public abstract class ReflectionUtil {
 		Class<?> clazz;
 		try {
 			clazz = Class.forName(className);
-			return clazz.newInstance();
+			return newInstance(clazz);
 		} catch (ClassNotFoundException e) {
-			throw new RuntimeException(e);
-		} catch (InstantiationException e) {
-			throw new RuntimeException(e);
-		} catch (IllegalAccessException e) {
 			throw new RuntimeException(e);
 		}
 	}
@@ -242,10 +240,37 @@ public abstract class ReflectionUtil {
 	 * @return the t
 	 */
 	public static <T> T newInstance(Class<T> clazz) {
-		try {
-			return clazz.newInstance();
-		} catch (Exception e) {
-			throw new RuntimeException(e);
+		Constructor<T> ctor = null;
+
+		if (classConstructorCache.containsKey(clazz)) {
+			ctor = (Constructor<T>) classConstructorCache.get(clazz);
+		} else {
+			// Try public ctor first
+			try {
+				ctor = clazz.getConstructor();
+			} catch (Exception e) {
+				// Swallow, no public ctor
+			}
+
+			// Next, try non-public ctor
+			try {
+				ctor = clazz.getDeclaredConstructor();
+				ctor.setAccessible(true);
+			} catch (Exception e) {
+				// Swallow, no non-public ctor
+			}
+
+			classConstructorCache.put(clazz, ctor);
+		}
+
+		if (ctor != null) {
+			try {
+				return ctor.newInstance();
+			} catch (Exception e) {
+				throw new RuntimeException("Cannot instantiate " + clazz, e);
+			}
+		} else {
+			throw new RuntimeException("No default constructor for " + clazz);
 		}
 	}
 
@@ -390,50 +415,8 @@ public abstract class ReflectionUtil {
 		Class<?> fieldClass = (Class<?>) type;
 
 		try {
-			if (fieldClass.equals(value.getClass())) {
-				// Types match
-				field.set(object, value);
-			} else {
-				// Types do not match, perform conversion where needed
-				if (fieldClass.equals(String.class)) {
-					field.set(object, value.toString());
-				} else if (fieldClass.equals(BigDecimal.class)) {
-					field.set(object, new BigDecimal(value.toString()));
-				} else if (fieldClass.equals(Date.class)) {
-					field.set(object, convertToDate(value));
-
-					// Use Number intermediary where possible
-				} else if (fieldClass.equals(Integer.class)) {
-					if (value instanceof Number) {
-						field.set(object, Integer.valueOf(((Number) value).intValue()));
-					} else {
-						field.set(object, Integer.valueOf(value.toString()));
-					}
-				} else if (fieldClass.equals(Long.class)) {
-					if (value instanceof Number) {
-						field.set(object, Long.valueOf(((Number) value).longValue()));
-					} else {
-						field.set(object, Long.valueOf(value.toString()));
-					}
-				} else if (fieldClass.equals(Double.class)) {
-					if (value instanceof Number) {
-						field.set(object, Double.valueOf(((Number) value).doubleValue()));
-					} else {
-						field.set(object, Double.valueOf(value.toString()));
-					}
-				} else if (fieldClass.equals(Float.class)) {
-					if (value instanceof Number) {
-						field.set(object, Float.valueOf(((Number) value).floatValue()));
-					} else {
-						field.set(object, Float.valueOf(value.toString()));
-					}
-
-					// Fallback to simply trying to set the field
-				} else {
-					field.set(object, value);
-				}
-			}
-
+			value = MappingUtil.convertValue(value, fieldClass);
+			field.set(object, value);
 		} catch (IllegalArgumentException e) {
 			Logger.error(ExceptionUtil.getStackTrace(e));
 		} catch (IllegalAccessException e) {
@@ -528,51 +511,6 @@ public abstract class ReflectionUtil {
 	 */
 	public static boolean isConcrete(Class<?> clazz) {
 		return !(isInterface(clazz) || isAbstract(clazz));
-	}
-
-	/**
-	 * Convert to date.
-	 * 
-	 * @param value
-	 *            the value
-	 * @return the date
-	 */
-	private static Date convertToDate(Object value) {
-		Date date = null;
-		if (value != null && !"".equals(value)) {
-			if (value instanceof Long) {
-				date = new Date(((Long) value).longValue());
-
-			} else if (value instanceof String) {
-				String val = (String) value;
-				int dateLength = String.valueOf(Long.MAX_VALUE).length();
-				if (dateLength == val.length()) {
-					date = new Date(Long.valueOf(val).longValue());
-				} else {
-					date = getDate(val);
-				}
-			} else {
-				date = (Date) value;
-			}
-		}
-		return date;
-	}
-
-	/**
-	 * Gets the date.
-	 * 
-	 * @param val
-	 *            the val
-	 * @return the date
-	 */
-	private static Date getDate(String val) {
-		try {
-			// Use ES internal converter
-			return XContentBuilder.defaultDatePrinter.parseDateTime(val).toDate();
-		} catch (Throwable t) {
-			Logger.error(ExceptionUtil.getStackTrace(t), val);
-		}
-		return null;
 	}
 
 	/**

